src/safeposix/syscalls/net_calls.rs

```diff
@@ -538,7 +538,85 @@ impl Cage {
+                    //so we just go to the match statement below -- "match how"
+                },
+                _ => {
+                    return syscall_error(Errno::ENOTSOCK, "netshutdown", "file descriptor is not a socket");
+                }
+            }
+        } else {
+            return syscall_error(Errno::EBADF, "netshutdown", "invalid file descriptor");
+        }
+
+        match how {
+            SHUT_RD => {
+                return syscall_error(Errno::EOPNOTSUPP, "netshutdown", "partial shutdown read is not implemented");
+            }
+            SHUT_WR => {
+                return Self::_cleanup_socket(self, fd, true, Some(&mut fdtable));
+            }
+            SHUT_RDWR => {
+                //BUG:: need to check for duplicate entries
+                return Self::_cleanup_socket(self, fd, false, Some(&mut fdtable));
+            }
+            _ => {
+                //See http://linux.die.net/man/2/shutdown for nuance to this error
+                return syscall_error(Errno::EINVAL, "netshutdown", "the shutdown how argument passed is not supported");
+            }
+        }
+    }
+
+    pub fn _cleanup_socket(&self, fd: i32, partial: bool, fdtable_lock: Option<&mut FdTable>) -> i32 {
+
+        //pass the lock of the FdTable to this helper. If passed table is none, then create new lock instance
+        let mut writer;
+        let fdtable = if let Some(rl) = fdtable_lock {rl} else {
+            writer = self.filedescriptortable.write().unwrap(); 
+            &mut writer
+        };
+
+        let mut mutmetadata = NET_METADATA.write().unwrap();
+        if let Some(wrappedfd) = fdtable.get(&fd) {
+            let mut filedesc = wrappedfd.write().unwrap();
+            match &mut *filedesc {
+                Socket(sockfdobj) => {
+                    let objectid = &sockfdobj.socketobjectid.unwrap();
+                    let localaddr = sockfdobj.localaddr.as_ref().unwrap().clone();
+                    let release_ret_val = mutmetadata._release_localport(localaddr.addr(), localaddr.port(), sockfdobj.protocol, sockfdobj.domain);
+                    if let Err(e) = release_ret_val {return e;}
+                    if !partial {
+                        mutmetadata.socket_object_table.remove(objectid);
+                        sockfdobj.state = ConnState::NOTCONNECTED;
+                    }
+                }
+                _ => {
+                    return syscall_error(Errno::ENOTSOCK, "listen", "file descriptor refers to something other than a socket");
+                }
+            }
+        } else {
+            return syscall_error(Errno::EBADF, "listen", "invalid file descriptor");
+        }
+
+        //have to take this out of the match because the fdtable already has a mutable borrow
+        //which means that I can't change it with a remove until after the fdtable mutable borrow is finished from the match statement
+        if !partial {
+            fdtable.remove(&fd); 
+        }
```

Comment:
If you put this at the end of the if block in the match it *should* be able to tell that the other mutable borrow is no longer needed. Try it out, and if it doesn't work keep it this way.

Reply:
It didn't work. I think that we have to keep it like this.